### PR TITLE
#1285 Make sure 'Source' and 'Raw' checkpoints are present when Standardization starts

### PR DESCRIPTION
**Pull request by AdrianOlosutean**
Ensure raw and source checkpoints are present in Standardization

### CODE REVIEW

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 1]
```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
...
+  /**
+   * Gets the number of records in source data by traversing Atum's checkpoints.
+   *
+   * @return The number of records in a checkpoint corresponding to source data (if available)
+   */
+  private def getSourceRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("source")
+  }
+
+  private def getFieldRawRecordFromCheckpoints(field: String) = {
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:465]: This method can encounter 3 types of error conditions. But It seems it distinguishes only 2 of them.
The 3 error conditions are:
- No checkpoint that has the name `name`
- There is a checkpoint `name`, but it doesn't have a `count` measurement
- There is a checkpoint `name`, it has a `count` measurement, but `count<0`

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 2]
```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
...
-    )
+    val optionResult = for {
+      checkpoint <- Atum.getControlMeasure
+        .checkpoints
+        .find(c => c.name.equalsIgnoreCase(field) || c.workflowName.equalsIgnoreCase(field))
+
+      measurement <- checkpoint.controls.find(m => m.controlType.equalsIgnoreCase(controlTypeRecordCount))
+      result <- Try({
+        val rowCount = measurement.controlValue.toString.toLong
+        if (rowCount >= 0) rowCount else throw new Exception("negative rowcount")
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:475]: Why not return `Left` here as well? `Option` might become unnecessary.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 3]
```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
...
 
-    val rawRecordCount = checkpointRawRecordCount match {
-      case Some(num) => num
-      case None      => df.count
+    (checkpointRawRecordCount, checkpointSourceRecordCount) match {
+      case (Right(x), Right(y)) => {
+        Atum.setAdditionalInfo(s"raw_record_count" -> x.toString)
+        Atum.setAdditionalInfo(s"source_record_count" -> y.toString)
+      }
+      case (Right(_), Left(er)) => throw new ValidationException("Field errors", Seq(er))
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:441]:
```suggestion
      case (Right(_), Left(er)) => throw new ValidationException("Checkpoint validation failed:", Seq(er))
```

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 4]
```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
...
   }
 
   /**
     * Gets the number of records in raw data by traversing Atum's checkpoints.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def getRawRecordCountFromCheckpoints: Option[Long] = {
+  private def getRawRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("row")
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:453]:
```suggestion
    getFieldRawRecordFromCheckpoints("raw")
```

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 5]
```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
...
+  }
+
+  /**
+   * Gets the number of records in source data by traversing Atum's checkpoints.
+   *
+   * @return The number of records in a checkpoint corresponding to source data (if available)
+   */
+  private def getSourceRecordCountFromCheckpoints: Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints("source")
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:462]: It could be useful to extract "Source" and "Raw" as constants.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 6]
```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
...
-        } else {
-          None
-        }
-      }
-      catch {
-        case NonFatal(_) => None
-      }
-    )
+    val optionResult = for {
+      checkpoint <- Atum.getControlMeasure
```
Comment (yruslan) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:468]: I know this is not part of your code, but it can be made better.
If `getSourceRecordCountFromCheckpoints` gets a `ControlMeasure` instance, it can be unit tested for various cases. `Atum.getControlMeasure` can be called outsode the method.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 2 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:475]: I think I did that because I could not use Either in the for comprehension, but I may be wrong

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 6 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:468]: Makes sense

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 7]
```diff
@@ -424,53 +425,58 @@ object StandardizationJob {
...
+        .find(c => c.name.equalsIgnoreCase(field) || c.workflowName.equalsIgnoreCase(field))
+
+      measurement <- checkpoint.controls.find(m => m.controlType.equalsIgnoreCase(controlTypeRecordCount))
+      result <- Try({
+        val rowCount = measurement.controlValue.toString.toLong
+        if (rowCount >= 0) rowCount else throw new Exception("negative rowcount")
+      }).toOption
+    } yield result
+
+    if (optionResult.isDefined) Right(optionResult.get) else Left(s"Invalid field ${field}")
```
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:479]: Curiously this was the shortest and accessible to use

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 1 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:465]: For this kind of use case, the Validated class would be quite a good fit

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 2 above)
Comment (yruslan) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:475]: Yep, you a correct. I've missed that, sorry.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 8]
```diff
@@ -424,53 +429,75 @@ object StandardizationJob {
...
-    * If raw record count is not available in checkpoints the method will calculate that count
-    * based on the provided raw dataframe.
+    * Adds metadata about the number of records in raw and source data by checking Atum's checkpoints first.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def addRawRecordCountToMetadata(df: DataFrame): Unit = {
-    val checkpointRawRecordCount = getRawRecordCountFromCheckpoints
+  def addRawAndSourceRecordCountsToMetadata: Unit = {
+    val checkpoints: immutable.Seq[Checkpoint] = Atum.getControlMeasure.checkpoints
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:437]: Why `immutable.`? (I believe it's the default)

File: spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala [hunk 9]
```diff
@@ -0,0 +1,63 @@
+package za.co.absa.enceladus.standardization
+
+import org.scalatest.FunSuite
+import za.co.absa.atum.model.{Checkpoint, Measurement}
+
+class StandardizationValidationSuite extends FunSuite {
```
Comment (benedeki) [inline spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala:6]: Not sure I like this unit test is not clear what it bounds to. 
For discussion: what about extracting the atum related logic to a dedicated `object` (or `class`), make it easier to test. Also `StandardizationJob` is becoming little fatty of a mixture of unrelated code.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 10]
```diff
@@ -102,7 +107,7 @@ object StandardizationJob {
     Atum.setAdditionalInfo(Constants.InfoVersionColumn -> reportVersion.toString)
 
     // Add the raw format of the input file(s) to Atum's metadta as well
-    Atum.setAdditionalInfo("raw_format" -> cmd.rawFormat)
+    Atum.setAdditionalInfo(rawFieldName + "_format" -> cmd.rawFormat)
```
[the remark this answers is not in the document]
Comment (yruslan) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:110]: It's odd, but for a different reason. We don't have a notion of 'source_format'. The row format is the format the data landed to Hadoop (e.g. CSV, XML, JSON). I'd leave the field name hardcoded.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 11]
```diff
@@ -424,53 +429,75 @@ object StandardizationJob {
...
+  /**
+   * Gets the number of records in source data by traversing Atum's checkpoints.
+   *
+   * @return The number of records in a checkpoint corresponding to source data (if available)
+   */
+  def getSourceRecordCountFromCheckpoints(checkpoints: immutable.Seq[Checkpoint]): Either[String, Long] = {
+    getFieldRawRecordFromCheckpoints(sourceFieldName, checkpoints)
+  }
+
+  private def getFieldRawRecordFromCheckpoints(checkpointName: String, checkpoints: immutable.Seq[Checkpoint]) = {
```
[the remark this answers is not in the document]
Comment (yruslan) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:473]: I think it is to collect all errors in `Left` and then throw an exception specifying the list of all errors. @AdrianOlosutean to confirm.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 12]
```diff
@@ -424,53 +429,75 @@ object StandardizationJob {
...
+
+    measurement match {
+      case Right(m) =>
+        val trial = Try({
+          val rowCount = m.controlValue.toString.toLong
+          if (rowCount >= 0) rowCount else throw new Exception(s"Negative value")
+        })
+        trial match {
+        case Success(value) => Right(value)
+        case Failure(ex) => Left(s"Wrong $checkpointName $controlTypeRecordCount value: ${ex.getMessage}")
```
[the remark this answers is not in the document]
Comment (yruslan) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:496]: We need `Long` to sanity check for negative values. Maybe we can workaround bogus values leading to the throw of `.toLong` by extracting a `safeToLong()` method that throws a more specific error? What do you think, guys?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 10 above)
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:110]: It was extracted by the IDE when I extracted the constant, I did not have a particular reason for that

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 13]
```diff
@@ -424,53 +429,75 @@ object StandardizationJob {
...
-    Atum.setAdditionalInfo(s"raw_record_count" -> rawRecordCount.toString)
   }
 
   /**
     * Gets the number of records in raw data by traversing Atum's checkpoints.
     *
     * @return The number of records in a checkpoint corresponding to raw data (if available)
     */
-  private def getRawRecordCountFromCheckpoints: Option[Long] = {
+  def getRawRecordCountFromCheckpoints(checkpoints: immutable.Seq[Checkpoint]): Either[String, Long] = {
```
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:460]: True, I'll remove them

File: spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala (hunk 9 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala:6]: I like this test suite and I like the idea of putting Standardization prerequisites validation into a separate object. Since
* It might be used in Conformance as well, for instance for jobs that do not require Standardization.
* It is expected that we will add more validations in the near future

When validations are extracted into a separate class or object, this test suite will fit perfectly.

@AdrianOlosutean , sorry that these ideas were not described in the GitHub issue. My bad, I didn't think it through properly.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:473]: Yes, you want to put those errors in the ValidationException further on, but the method will have just one error at a time. @benedeki , do you mean to have a for comprehension with optionals?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala (hunk 9 above)
Comment (AdrianOlosutean) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala:6]: Ok, I will extract it into a separate object

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala [hunk 14]
```diff
@@ -424,53 +429,75 @@ object StandardizationJob {
...
+    val checkpointRawRecordCount = getRawRecordCountFromCheckpoints(checkpoints)
+    val checkpointSourceRecordCount = getSourceRecordCountFromCheckpoints(checkpoints)
+
+    val errorMessage = "Checkpoint validation failed:"
 
-    val rawRecordCount = checkpointRawRecordCount match {
-      case Some(num) => num
-      case None      => df.count
+    (checkpointRawRecordCount, checkpointSourceRecordCount) match {
+      case (Right(x), Right(y)) => {
```
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:445]: Actually here it's needed cause the code inside is imperative. I should restructure it

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:496]: Where would this method be located?

File: spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala (hunk 9 above)
Comment (yruslan) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala:6]: `ControlInfoValidation` maybe?

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (yruslan) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:496]: It's not implemented, but we can put it into `Utils`.

File: spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala (hunk 9 above)
Comment (benedeki) [reply spark-jobs/src/test/scala/za/co/absa/enceladus/standardization/StandardizationValidationSuite.scala:6]: For the record, I don't have any issue with the suite itself. 😄 
It's just it's not clear what it unit tests. (e.g. the `Shift`+`Ctrl`+`T` key combination doesn't tap here

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:445]: Not sure why this make them needed and what does it change? 🤔

[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:445]: The Atum.setAdditionalInfo lines are side effects

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 11 above)
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:473]: Indeed, something like that I was thinking to do, but I thought I could not use Try properly in the for comprehension. Actually I can

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 14 above)
[the remark this answers is not in the document]
Comment (benedeki) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:445]: Yup. But the Atum functions are of same return type as this one here (`Unit`). And if not I don't think the curly braces would fix it. 😉 
So I don't think they are needed.

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala (hunk 12 above)
[the remark this answers is not in the document]
Comment (AdrianOlosutean) [reply spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/StandardizationJob.scala:496]: I wouldn't say it's criptic. As you can see in the testing suite, for an empty string value there it would something like : Wrong raw $controlTypeRecordCount value: For input string ""

File: spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+      case (Success(x), Success(y)) => FieldCounts(x, y)
+      case (Success(_), Failure(er)) => throw new ValidationException(errorMessage, Seq(er.getMessage))
+      case (Failure(er), Success(_)) => throw new ValidationException(errorMessage, Seq(er.getMessage))
+      case (Failure(er1), Failure(er2)) =>
+        throw new ValidationException(errorMessage, Seq(er1.getMessage, er2.getMessage))
+    }
+  }
+
+
+  def getFieldRecordFromCheckpoints(checkpointName: String, checkpoints: Seq[Checkpoint]): Try[Long] = {
```
Comment (benedeki) [inline spark-jobs/src/main/scala/za/co/absa/enceladus/standardization/ControlInfoValidation.scala:54]:
```suggestion
  def getFieldCountFromCheckpoint(checkpointName: String, checkpoints: Seq[Checkpoint]): Try[Long] = {
```
The function is public and gets a particular field from the checkpoint, but the name is very general. This is just a suggestion tho. There might easily be better names. 😉
